### PR TITLE
*: fix a data race for SelectForUpdate executor

### PR DESCRIPTION
**Pull request by tiancaiamao**
`select * from (select * from t for update) t join t1 for update`

There will be two SelectForUpdate executors, both modify TxnCtx in NextChunk

```
2018/03/20 14:49:53.005 adapter.go:308: [debug] [230][ActivePendingTxn] begin
2018/03/20 14:49:53.005 adapter.go:376: [debug] query succ=true txnStartTS=398859591484702720 database=test costTime=183.993µs sql=begin type=query connectionId=230
2018/03/20 14:49:53.006 compiler.go:78: [warning] [EXPENSIVE_QUERY] select * from (select * from t for update) t join t1 for update
2018/03/20 14:49:53.006 adapter.go:308: [debug] [230][ActivePendingTxn] select * from (select * from t for update) t join t1 for update
==================
WARNING: DATA RACE
Write at 0x00c4222d99e0 by goroutine 221:
  github.com/pingcap/tidb/executor.(*SelectLockExec).NextChunk()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:496 +0x232
  github.com/pingcap/tidb/executor.(*ProjectionExec).NextChunk()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:648 +0xf7
  github.com/pingcap/tidb/executor.(*recordSet).NextChunk()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:111 +0x95
  github.com/pingcap/tidb/session.GetRows4Test()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/tidb.go:225 +0x288
  github.com/pingcap/tidb/util/testkit.(*TestKit).ResultSetToResult()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:179 +0x142
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustQuery()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:173 +0x4d3
  github.com/pingcap/tidb/executor_test.(*testSuite).TestSelectForUpdate()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor_test.go:1973 +0xbb3
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:130 +0x133
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:159 +0x91
  github.com/pingcap/tidb/executor_test.(*testSuite).TestSelectForUpdate()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor_test.go:1927 +0x55f
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:130 +0x133
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:159 +0x91
  github.com/pingcap/tidb/executor_test.(*testSuite).TestSelectForUpdate()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor_test.go:1922 +0x483
  github.com/pingcap/tidb/executor_test.(*testSuite).TestSelectForUpdate()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor_test.go:1919 +0x397
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:573 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:308 +0xc0
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkTest.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:798 +0x954
  github.com/pingcap/tidb/vendor/github.com/pingcap/check.(*suiteRunner).forkCall.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/vendor/github.com/pingcap/check/check.go:692 +0x89

Previous write at 0x00c4222d99e0 by goroutine 322:
  github.com/pingcap/tidb/executor.(*SelectLockExec).NextChunk()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:496 +0x232
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchOuterChunks()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:212 +0x283
```

@coocood @iamxy @zz-jason

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -93,7 +93,10 @@ type TransactionContext struct {
 	SchemaVersion int64
 	StartTS       uint64
 	Shard         *int64
-	TableDeltaMap map[int64]TableDelta
```
Comment (XuHuaiyu) [inline sessionctx/variable/session.go]: Any other attributes of TransactionContext should be moved into Mu?

**Comment by coocood** [comment]
@tiancaiamao 
We can `UpdateDeltaForTable` in `Open` and avoid the mutex.

File: sessionctx/variable/session.go (hunk 1 above)
Comment (tiancaiamao) [reply sessionctx/variable/session.go]: Keep a small lock scope makes better performance.
If we found data race on them, I would update.

**Comment by tiancaiamao** [comment]
@coocood @lamxTyler PTAL

**Pull request merged**